taichi/transforms/demote_operations.cpp

```diff
@@ -14,6 +14,37 @@ class DemoteOperations : public BasicStmtVisitor {
   DemoteOperations() : BasicStmtVisitor() {
   }
 
+  void visit(BitExtractStmt *stmt) override {
+    // @ti.func
+    // def bit_extract(input, begin, end):
+    //   return (input >> begin) & ((1 << (end - begin)) - 1)
+
+    auto one = Stmt::make<ConstStmt>(LaneAttribute<TypedConstant>(1));
+    auto begin =
+        Stmt::make<ConstStmt>(LaneAttribute<TypedConstant>(stmt->bit_begin));
+    auto esbeg = Stmt::make<ConstStmt>(
+        LaneAttribute<TypedConstant>(stmt->bit_end - stmt->bit_begin));
+    auto input_sar_begin = Stmt::make<BinaryOpStmt>(BinaryOpType::bit_sar,
+                                                    stmt->input, begin.get());
+    auto one_shl_esbeg =
+        Stmt::make<BinaryOpStmt>(BinaryOpType::bit_shl, one.get(), esbeg.get());
+    auto one_shl_esbeg_sub_one = Stmt::make<BinaryOpStmt>(
+        BinaryOpType::sub, one_shl_esbeg.get(), one.get());
+    auto ret =
+        Stmt::make<BinaryOpStmt>(BinaryOpType::bit_and, input_sar_begin.get(),
+                                 one_shl_esbeg_sub_one.get());
+
+    stmt->replace_with(ret.get());
+    modifier.insert_before(stmt, std::move(one));
+    modifier.insert_before(stmt, std::move(begin));
+    modifier.insert_before(stmt, std::move(esbeg));
+    modifier.insert_before(stmt, std::move(input_sar_begin));
+    modifier.insert_before(stmt, std::move(one_shl_esbeg));
+    modifier.insert_before(stmt, std::move(one_shl_esbeg_sub_one));
+    modifier.insert_before(stmt, std::move(ret));
```

Comment:
Actually you can make use of `DelayedIRModifier::insert_before(Stmt *old_statement, VecStatement &&new_statements)`  and `VecStatement::push_back(Args &&... args)` instead of `Stmt::make` to simplify this part.

Reply:
But I think keep getting the `back()` for operands is inconvenient here, for example, `one_shl_esbeg_sub_one`.

Reply:
It returns `back()`, for example, you can write `auto one_shl_esbeg_sub_one = VecStatement::push_back( BinaryOpType::sub, one_shl_esbeg, one);`

Reply:
Great! It's really convenient then.